src/App.js

```diff
@@ -29,37 +29,108 @@ const App = () => {
   // This starts state off as a 2D array of JS objects with
   // empty value and unique ids.
   const [squares, setSquares] = useState(generateSquares());
+  const [playerTurn, setPlayerTurn] = useState(PLAYER_1);
+  const [won, setWon] = useState(null);
 
   // Wave 2
   // You will need to create a method to change the square
   //   When it is clicked on.
   //   Then pass it into the squares as a callback
 
+  const updateBoard = markedSquare => {
+    
+    const newBoard = [...squares];
+    for (let row of newBoard) {
+      for (let square of row) {
+        if (square.id === markedSquare.id) {
+          square.value = markedSquare.value;
+        }
+      }
+    }
+    setSquares(newBoard);
+    let result = checkForWinner();
+    decideContinue(result);
+  };
+
+  const switchPlayers = () => {
+    if (playerTurn === PLAYER_1) {
+      setPlayerTurn(PLAYER_2);
+    } else if (playerTurn === PLAYER_2) {
+      setPlayerTurn(PLAYER_1);
+    }
+  };
+
+  const decideContinue = (winner) => {
+    if (!winner) { 
+      switchPlayers();
+    } else if (winner) {
+      setWon(winner);
+    }
+  };
```

Comment:
We can further refactor this code with the following:

1. Keeping `won` as a normal `let` variable rather than a state variable. 
2.  Eliminating `decideContinue()` and use a ternary to determine the current player to be passed into `setPlayerTurn()`.
3. Moving`checkForWinner()` outside of the click handler so it can be called in each re-render. (in a comment after the definition for `checkForWinner()`.

```suggestion


  // Wave 2
  // You will need to create a method to change the square
  //   When it is clicked on.
  //   Then pass it into the squares as a callback

  const updateBoard = markedSquare => {
    
    const newBoard = [...squares];
    for (let row of newBoard) {
      for (let square of row) {
        if (square.id === markedSquare.id) {
          square.value = markedSquare.value;
        }
      }
    }
    setSquares(newBoard);
    setPlayerTurn(playerTurn === PLAYER_1 ? PLAYER_2 : PLAYER_1);
  };

```